Fix Buildkite test collector support when tests execution fail

wzieba opened the pull request:
Closes: AINFRA-258

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR fixes unit tests collection in cases when unit tests fail.

#### Issue

I spotted, that `jacocoTestReport` doesn't reuse output of `:WooCommerce:testJalapenoDebugUnitTest` task. Take a random build from current pull requests, e.g:

https://buildkite.com/automattic/woocommerce-android/builds/28754/steps?jid=01966d6b-e9a2-49c3-adb8-ac77b8c0e369#01966d6b-e9a2-49c3-adb8-ac77b8c0e369/357-373

and see, that `:WooCommerce:testJalapenoDebugUnitTest` was executed when running `jacocoTestReport`, but just before, it was already executed in previous step. Looking at Gradle Scan for a reason why it was executed, one can spot

<img width="1138" alt="Screenshot 2025-04-29 at 13 27 24" src="https://github.com/user-attachments/assets/5e4cee8f-0740-49c8-abae-2ece9cb2a6d4" />

I don't know which operation exactly breaks `stableClasspath`, but I also don't think it's crucial to understand it at this moment.

This is problematic for several reasons:
- **not working Buildkite Test Collector collection when tests fail** 
In such case, we don't report it back to Buildkite as there's no `set +e`/`set -e` for `jacocoTestReport`, so this task failing due to tests failure, simply exists the job ([source](https://buildkite.com/automattic/woocommerce-android/builds/28794/steps?jid=01968140-dd82-4c02-819e-34f1b818b541), no `Tests` tab).
- unnecessary **longer job time** (there's no reason to run tests twice)
- **unpredicted behavior**: we run tests, modify build outputs and then run tests again. This results with behavior like:
  - tests are passing for the first time, but not for the second time ([source](https://buildkite.com/automattic/woocommerce-android/builds/28776/steps?jid=01967d1d-d029-4ebf-8995-e9b92312f1dd#01967d1d-d029-4ebf-8995-e9b92312f1dd/354-370))
  - different tests are failing on first and second run ([source](https://buildkite.com/automattic/woocommerce-android/builds/28794/steps?jid=01968140-dd82-4c02-819e-34f1b818b541))

#### Solution

Run `jacocoTestReport` as one of the tasks during tests tasks execution. 
- this will allow the whole script to finish, even if `jacocoTestReport` fails
- this also runs `jacocoTestReport` before any other operations related to reporting, making sure, that the build can safely reuse output of previous tasks.

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: https://woomobilep2.wordpress.com/2024/05/06/woocommerce-mobile-quality-report-march-april/#comment-12036 -->

- For "**unpredicted behavior**" and "unnecessary **longer job time**" see [build](https://buildkite.com/automattic/woocommerce-android/builds/28800/steps?jid=0196815e-1d87-4db3-9d06-8abe41bbb621) of #13982 . After cherry-picking commits from this branch, unit tests are just fine.
- For  "**not working Buildkite Test Collector collection** " see: [build](https://buildkite.com/automattic/woocommerce-android/builds/28812/steps?jid=019681a3-565a-4f0c-a4fa-8dbeade308f4) of #13988. The tests _are_ collected even that the tests task failed.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
<img width="814" alt="Screenshot 2025-04-29 at 16 44 14" src="https://github.com/user-attachments/assets/e4b8c390-0b7c-449d-8b9f-6175d6e89aa4" />


- [ ] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on big (tablet) and small (phone) in case of UI changes, and no regressions are added.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

wzieba committed:
Fix missing buildkite test collector support when JaCoCo fails

In case when `jacocoTestReport` task fails, our script was not moving test results to the correct directory, as it was exiting on error before collection.

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

wzieba committed:
Run `jacocoTestReport` on the same Gradle invocation. Upload code coverage only on success tests

This will reduce chances of `testJalapenoDebugUnitTest` task being not up to date when running `jacocoTestReport`, which caused running unit tests again, without any reason.

wzieba committed:
Add missing modules test tasks to JacocoReport dependencies

To address issues like: Reason: Task ':jacocoTestReport' uses this output of task ':libs:apifaker:testDebugUnitTest' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.

wzieba committed:
Print "Uploading code coverage" in a new line

wzieba committed:
Open `Testing` by default

There are two reasons for this:
- with the current setup, we were opening "Copying test logs for test collector" if tests failed, because this was the last collapsed regular group (docs: https://buildkite.com/docs/pipelines/configure/managing-log-output#collapsing-output) and it was confusing: tests did fail, not copying test logs
- developers are much more interested in details of `Testing` step than (empty) details of "Copying test logs" so it makes sense to open former by default

wzieba commented on .buildkite/commands/run-unit-tests.sh:
```diff
@@ -8,19 +8,17 @@ install_gems
 bundle exec fastlane run configure_apply
 
-echo "--- 🧪 Testing"
+echo "+++ 🧪 Testing"
```
To add more context to the commit message ( ca437a01b5413ac0357f60b4bf23ab68a47ac678 ) , here are the screenshots:

| Before | After | 
| --- | --- | 
| <img width="982" alt="Screenshot 2025-04-29 at 14 27 10" src="https://github.com/user-attachments/assets/fa54cfb4-ddd4-4748-af99-eb88e2dd6d25" /> | <img width="1248" alt="Screenshot 2025-04-29 at 15 46 55" src="https://github.com/user-attachments/assets/0e1c54c8-cbb7-449d-a053-b5fb3b0f5b7e" /> |

ParaskP7 replied:
Praise (❤️): That's a really nice touch! 🥇

wzieba committed:
echo report tests status group in a new line

wzieba committed:
Actually print some section headers in a new line

ParaskP7 commented on config/gradle/jacoco.gradle:
```diff
@@ -16,6 +16,8 @@ rootProject.afterEvaluate {
                 ':libs:fluxc-tests:testDebugUnitTest',
                 ':libs:login:testDebugUnitTest',
+                ':libs:apifaker:testDebugUnitTest',
+                ':libs:commons:testDebugUnitTest',
```
Question (❓): This `libs:commons` module doesn't have any tests, is adding it here going to help? Or maybe you did that to prepare for when that module will add its first test? 🤔

PS: How about other modules that also test-less, like `libs:fluxc-annotations`, `:libs:fluxc-processor` and `:libs:detektrules`?

wzieba replied:
Good idea to keep all modules here, but these modules don't have a test task (you can check with `./gradlew :<module>:tasks --all`). They are all projects with JVM plugin applied, so if they don't have test sources, tasks are not available.

I added `libs:commons` because without it Gradle was reporting hidden dependency issue:

> Reason: Task ':jacocoTestReport' uses this output of task ':libs:commons:testDebugUnitTest' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.

ParaskP7 reviewed (approved):
👋 @wzieba !

I have reviewed and tested this PR as per the instructions, everything works as expected, nice work on that! 🌟 

> I spotted, that jacocoTestReport doesn't reuse output of :WooCommerce:testJalapenoDebugUnitTest task.

Good catch, I too wonder now, why... 🤔 

> ...but I also don't think it's crucial to understand it at this moment.

I agree. 👍 

-----

I have left just one question (❓)for you. I am going to approve this PR anyway, since none is blocking. I am **NOT** going to merge this PR yet to give you some time to apply any of my suggestions. However, feel free to ignore them and merge the PR yourself.
